src/transformers/models/fast/modeling_fast.py

```diff
@@ -0,0 +1,932 @@
+        embs = hidden[:, 1:, :, :]  # 4*4*640*640
+
+        selected_masks = ohem_batch(texts, gt_texts, training_masks)
+        loss_text = dice_loss_with_masks(texts, gt_texts, selected_masks, reduce=False)
+
+        selected_masks = gt_texts * training_masks
+        loss_kernel = dice_loss_with_masks(kernels, gt_kernels, selected_masks, reduce=False)
+        loss_kernel = torch.mean(loss_kernel, dim=0)
+
+        loss_emb = emb_loss_batch(embs, gt_instances, gt_kernels, training_masks, reduce=False, bg_sample=self.loss_bg)
+
+        return torch.mean(loss_text) + torch.mean(loss_kernel) + torch.mean(loss_emb)
+
+    @add_start_docstrings_to_model_forward(FAST_FOR_CAPTIONING_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=FASTForImageCaptioningOutput, config_class=_CONFIG_FOR_DOC)
+    def forward(
+        self,
+        pixel_values: torch.FloatTensor,
+        output_hidden_states: Optional[bool] = True,
+        return_dict: Optional[bool] = None,
+        labels: Dict = None,
+    ):
+        r"""
+        labels (`torch.LongTensor` of shape `(batch_size, height, width)`, *optional*):
+            Ground truth semantic segmentation maps for computing the loss. Indices should be in `[0, ...,
+            config.num_labels - 1]`. If `config.num_labels > 1`, a classification loss is computed (Cross-Entropy).
+
+        Returns:
+
+                Examples:
+
+        ```python
+        >>> from transformers import FastImageProcessor, FASTForImageCaptioning >>> from PIL import Image >>> import
+        requests >>> url = "https://huggingface.co/datasets/Raghavan/fast_model_samples/resolve/main/img657.jpg" >>>
+        image = Image.open(requests.get(url, stream=True).raw).convert("RGB") >>> processor =
+        FastImageProcessor.from_pretrained("Raghavan/fast_base_tt_800_finetune_ic17mlt") >>> model =
+        FASTForImageCaptioning.from_pretrained("Raghavan/fast_base_tt_800_finetune_ic17mlt") >>> inputs =
+        processor(image, return_tensor="np") >>> # forward pass >>> outputs =
+        model(pixel_values=torch.tensor(inputs["pixel_values"])) >>> target_sizes = [(image.shape[1], image.shape[2])
+        for image in inputs["pixel_values"]] >>> text_locations = processor.post_process_text_detection(outputs,
+        target_sizes) >>> print(text_locations[0]["bboxes"][0][:10]) [484, 175, 484, 178, 483, 179, 452, 179, 452, 182]"""
+        # outputs = {}
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+        hidden_states = self.backbone(pixel_values)
+
+        hidden_states = self.neck(hidden_states)
+
+        text_detection_output = self.det_head(hidden_states)
+
+        loss = None
+        if labels:
+            out = self._upsample(text_detection_output, pixel_values.size(), scale=1)
+            loss = self.loss(out, labels)
+        text_detection_output = self._upsample(text_detection_output, pixel_values.size(), scale=4)
```

Comment:
what does the model output? Should this be logits?

Reply:
The model ouputs hidden_state with same shape as input image. This hidden_state is then used find text in image.